### DISCUSSION TITLE
Updates to the Code of Conduct

### DISCUSSION THREAD

**Discussion by cblecker** [cncf/wg-coc#16] [Request for Feedback]
The working group is seeking feedback on the current CNCF Code of Conduct that will be enforced by both the CNCF Code of Conduct Committee this working group will be creating, as well as by project Code of Conduct bodies.

The current Code of Conduct in force is available here: https://github.com/cncf/foundation/blob/main/code-of-conduct.md

A working draft of the Code of Conduct has been copied to this repo, but as of the time of the start of this discussion, it is word for word identical to the in force CoC: https://github.com/cncf/wg-coc/blob/main/code-of-conduct-proposed.md

If you have feedback on this document, please leave your comments in this discussion. Once consensus around points is reached, then PRs will be created to update the draft.

**Comment by natalisucks** [comment]
Thanks for starting this thread @cblecker. This might be a small point, but the CoC starts with a "Contributor" CoC and names those adhering to it as "contributors and maintainers". However, the definition of contributor here isn't laid out, and could be interpreted as only those folks who are committing code/PRs, while newbies who have yet to join certain orgs or folks who haven't contributed code might not feel this applies to them.

I'd like to propose that the subheading "Contributor Code of Conduct" be changed to "Community Code of Conduct" (under the languages list), and that the qualifier "participants" be added alongside contributors and maintainers in the paragraph following. This way, we ensure that every single person interacting with the CNCF community, through projects, GitHub, Slack, etc, is accounted for in very purposeful wording of the CoC.

**Reply by joannalee333** [reply]
@natalisucks, thanks for your feedback. I will modify [my existing PR](https://github.com/cncf/wg-coc/pull/23) to include your suggested changes.

**Comment by riecatnor** [comment]
Hi @cblecker I am a bit new around the CNCF, though I have been hanging around in the coc-wg Slack for a while and was able to join the recent Code of Conduct related meetup at Kubecon (virtually). I have some feedback that I hope will be helpful for this process and the outcome. I am still getting to know the CNCF community and its culture, so my suggestions might be hit or miss. 

A bit of background for those who don’t know me yet: I worked on creating the Fedora Project’s current Code of Conduct, as well as managing CoC incidents for Fedora for 3+ years. More recently I have been working directly on CoC related work, developing further CoC processes for the Fedora community, consulting on CoC, and researching open source communities CoC’s and related processes/documentation. 

Based on my experience and research, my overall sense about the CNCF Code of Conduct in its current state is that it is a bit vague in certain areas (which might be intentional). I think there are some small additions/changes that would help provide more clarity, coverage, inclusivity, and understanding for community members and those stewarding the Code of Conduct and its processes. 

**Contributor Code of Conduct Section**
- Would recommend adding a catch all “or any other dimension of diversity”
- Would recommend alphabetizing the dimensions of diversity.
- If you want to clarify or be clear that specific groups are welcome in your organization, you might consider adding additional dimensions of diversity, such as: Ability, background, caste, education, ethnicity, family status, geographic location, marital status, national origin, native language, socioeconomic status, tribe, veteran status. See references below for examples for what is commonly, and less commonly, referenced. 

**Scope Section**
- Would consider making this a bit longer by adding some specifics related to the CNCF community. Many codes of conduct provide examples of community spaces, or event specifics after stating “including, but not limited to”. Keeping it vague seems like an okay approach, but by doing so, you leave it open to interpretation/argument. See references below for examples.  

**Our Standards Section**
- The unacceptable behavior list is short, and I think the last line about professional setting does a decent job at being a catch all. That being said, you might find that people try to skirt around accountability for their behavior by pointing to this list and saying “well it doesn’t say I can’t do xyz there”. Based on references below and CNCF culture, you might consider making additions such as “unwelcome physical contact”, “violence or threats of violence”, “Sexist, racist, homophobic, transphobic, ableist, or exclusionary statements, even if they were meant as jokes”, “sustained disruption of online discussions or presentations (virtual or in person)”, or “Initiating controversy for controversy’s sake”. See references below for more examples. 
- Building off the previous comment on this thread from @natalisucks (which I think is a great point and should be applied through out the Code and other docs as applicable), you might consider changing “Project maintainers” to “Project maintainers, community members, and moderators” in the last section unless the only folks with moderation privileges across all platforms and community spaces would be maintainers. 
- Would make a small change to this sentence in the last paragraph of this section: "By adopting this Code of Conduct, project maintainers commit themselves to fairly and consistently applying these principles _in alignment with to the Code of Conduct_ to every aspect of managing this project." From my experience, moderators/project maintainers are not exempt from disagreement with the spirit and/or material in a CoC. Its good to set an expectation that they will exercise their mod/maintainer responsibilities in alignment with the code (not their interpretation).

**Reporting Section**
- I need to review the corresponding documents before commenting on this section.

**Enforcement**
- Would recommend adding something to the extent of “Upon our review and investigation of the reported incident, we will determine what action is appropriate based on this Code and its associated documents.” I think this helps clarify the expectation that the Committee overseeing the Code has decision making power about what actions are taken.
- I need to review the corresponding documents before further commenting on this section.

References used, and if any changes are adopted, attribution should be made to some/all of these sources:
https://docs.fedoraproject.org/en-US/project/code-of-conduct/
https://www.python.org/psf/conduct/
https://wiki.gnome.org/Foundation/CodeOfConduct
https://www.mozilla.org/en-US/about/governance/policies/participation/

I hope this feedback is helpful!

**Reply by joannalee333** [reply]
@riecatnor, thanks for your input. I've modified an [existing PR](https://github.com/cncf/wg-coc/pull/23) to incorporate your suggestions.

**Comment by tpepper** [comment]
It might be informative for folks engaging in a potential CoC update to get some education on the topic, eg: https://www.strangertickets.com/events/132553120/consent-and-leadership-4-writing-codes-of-conduct

**Comment by tpepper** [comment]
In reviewing the latest incident handling proposal (https://github.com/cncf/wg-coc/blob/main/coc-incident-resolution-procedures-proposed.md as of https://github.com/cncf/wg-coc/commit/80fc613779e0304c3fd48fc0faeeb799505fd7ba), I noticed that it calls out retaliation as a CoC violation in https://github.com/cncf/wg-coc/blob/main/coc-incident-resolution-procedures-proposed.md#no-retaliation but while made explicit there it is not in the explicit examples of unacceptable behavior list in the proposed CoC.  Surely implied in "Other conduct which could reasonably be considered inappropriate in a professional setting", but perhaps worth calling out explicitly for consistency if it remains explicit in other CoCC documentation.

**Reply by joannalee333** [reply]
Thanks, @tpepper. I opened a PR to modify the CoC itself to include retaliation as a form of unacceptable behavior

**Comment by marinapistarini** [comment]
Hi Everyone. My name is Marina. I just sent this[ pull request](https://github.com/cncf/wg-coc/pull/22) as I'm interested in collaborating, and would love to know if these edits are useful. I would love to know if there is a specific section where I can focus on developing. 

Thanks in advance for the advice! 

Bests, 

Marina.

**Reply by joannalee333** [reply]
Hi @marinapistarini, thank you! I think your suggested changes are helpful. I'll comment more specifically on the PR you opened

**Comment by mattfarina** [comment]
The scope of the CoC has been updated to:

> This code of conduct applies both within project spaces and in other spaces (public or private) when an individual's words or actions directly impact a CNCF project or the CNCF community.

What does "directly impact" mean?

In the spirit testing, here are a couple cases and I wonder if the new CoC would apply to them:

1. On social media, someone who isn't representing themselves as a contributor or maintainer of a CNCF project but is one, makes a political or religious (belief statement) that some in the community don't like. This impacts the community by making some people upset. Is this a direct impact?
2. At a private get together (e.g., dinner get together in someone's home) a person expresses that they support a political candidate or topic that someone else at the get together strongly disagrees with. This causes tension between the two people so they don't want to work or collaborate with each other. For context, both people hold leadership positions on the same CNCF project. Is this directly impacting?

Note, both of these are situations are _based_ on real world situations I've encountered in open source or business.

**Reply by liggitt** [reply]
> The scope of the CoC has been updated to:
> 
> > This code of conduct applies both within project spaces and in other spaces (public or private) when an individual's words or actions directly impact a CNCF project or the CNCF community.

Has that been updated, or is it a proposed update? I don't see that in https://github.com/cncf/foundation/blob/main/code-of-conduct.md#scope or https://github.com/cncf/wg-coc/blob/main/code-of-conduct-proposed.md#scope

**Reply by mattfarina** [reply]
https://github.com/cncf/wg-coc/pull/23 is where you can find it

**Reply by dims** [reply]
@liggitt everything in this repository is essentially under debate, folks are welcome to file PR(s), update PR(s) and do all the normal stuff we do :)

**Reply by liggitt** [reply]
sounds good, I just wasn't sure where to look and the comment made it sound like the change had already gone into effect

**Reply by mattfarina** [reply]
@liggitt sorry I wasn't clear in my initial comment. That was my fault for not pointing out it's proposed.

**Reply by deads2k** [reply]
Even with the recent  update to 
> This code of conduct applies both within project spaces and in other spaces when an individual's words or actions directly pertain to a CNCF project or the CNCF community.

I think @mattfarina's scenarios are still interesting.  In his scenarios, do the opinions expressed "pertain to the community"?  If not, what is the basis for claiming that other words said in those scenarios (the harrassment example from the comments in #23) would pertain to the community?

**Reply by parispittman** [reply]
for #1: does the opinions expressed line up with community values and do they push people away? i'd argue that's important when you have a value that's explicitly around inclusion.

**Reply by endocrimes** [reply]
This is a pretty nuanced discussion that is pretty hard to get right over GitHub. To understand what would or would not be in scope somewhat involves an understanding of the _intent_ of the Code of Conduct.

To quote the current proposal (not from #23):

> As contributors and maintainers in the CNCF community, and in the interest of fostering an open and welcoming community, we pledge to respect all people who contribute through reporting issues, posting feature requests, updating documentation, submitting pull requests or patches, and other activities.
> 
> We are committed to making participation in the CNCF community a harassment-free experience for everyone, regardless of level of experience, gender, gender identity and expression, sexual orientation, disability, personal appearance, body size, race, ethnicity, age, religion, or nationality.

What does that mean?

The code of conduct has a few primary goals:
- Fostering an open and welcoming community
- Centers respecting people who contribute to the project in _any_ form
- With some explicit examples of some categories of folks that we wish to protect.
- And to not tolerate harassment of members of the community.

As @joannalee333 says in her PR, the change is being proposed with the following intent:

>     A community member is harassing another community member privately, outside of project spaces (e.g., stalking them, making unwanted sexual advances, threatening violence against them, etc.)
>     A community member engages in harassment of other community members in public spaces not controlled by the project (e.g., via Instagram or Twitter)

These are both things we've faced inside our communities including many current, and (sadly) former contributors have experienced in the past too. In the current code of conduct, it can be hard to take otherwise common sense actions without opening up whataboutism and risk to the committee and the project - as the boundary is _too_ unclear. We need to include these types of actions in order to keep our communities safe.

Building _inclusive_ environments always requires acknowledging that you need to _exclude_ some types of behavior, or you fall victim to the paradox of tolerance.

To map this back to your examples:

> On social media, someone who isn't representing themselves as a contributor or maintainer of a CNCF project but is one, makes a political or religious (belief statement) that some in the community don't like. This impacts the community by making some people upset. Is this a direct impact?

I'd unpack this phrasing a little, but that is _unlikely_ to be productive. I'm instead going to focus on _my_ rubric of how I'd consider this.

There's an important distinction to be made here between a statement that some _do not like_, vs one that _does harm to members of the community_ or _discourages people from being part of the community_. I think under the revised wording, that something that does the latter would potentially pertain to the community and its status.

My personal litmus test is "is someone in a position of power or majority punching down at an individual or a smaller minority group?" or "is someone attempting to intrude on the bodily autonomy of others?" - if the answer to either question is yes, they are quite likely contributing to an unsafe and unwelcoming environment.

The impact of those statements varies in a number of factors. Are they in a position of leadership? Are they often or usually identified as someone who is part of a project - even if they're not advertising it _right now_? If yes, then it probably has an outsized impact on how people feel safe contributing to the project, regardless of how they're currently presenting.

If both of these are true, or if either is _especially_ weighty, then the statement potentially falls under being a _potential_ violation of a code of conduct.

If a statement is relatively harmless "I like cheese" or "I'm voting for [The Monster Raving Loony Party](https://en.wikipedia.org/wiki/Official_Monster_Raving_Loony_Party)" - it's not going to be an issue.

But if that statement is "I don't think [group i am not a member of] should have the right to bodily autonomy", it probably falls under creating a hostile environment for other contributors. Especially if made by someone in a leadership position or is a long time prolific contributor.

For the second point:

> At a private get together (e.g., dinner get together in someone's home) a person expresses that they support a political candidate or topic that someone else at the get together strongly disagrees with. This causes tension between the two people so they don't want to work or collaborate with each other. For context, both people hold leadership positions on the same CNCF project. Is this directly impacting?

I'd apply the same rubric as the first. "Supporting a political candidate" _probably_ couldn't fall under this scope, as much as current political tensions and dividing lines in some countries easily could result in a breach of a code of conduct.

That said, if they made statements that punch down at a group they are not a member of (e.g racist/misogynistic/homophobic remarks), and they are at a dinner where any project related things are discussed - then it would probably form a code of conduct issue just as doing so with coworkers could become a HR issue.

**Reply by joannalee333** [reply]
Hi everyone! Great discussion here. This is in fact nuanced, and I want to make sure we get this right. We want the CoC scope to be broad enough to cover situations where there is harassment directed at specific community members (in any forum, public or private - not just project spaces), but not so broad that it enables the CoC to itself be used as a means for excluding people from the community due to matters in their private lives that are completely unrelated to the work of the community and aren't directed at community members. I will spend some time over the weekend reading and re-reading the comments carefully (so keep them coming!), and early next week, I will suggest further modifications/refinements to the language to strike the balance that I think we're all trying to achieve.

**Reply by mattfarina** [reply]
I hope we can all appreciate how difficult it is to have conversations about this topic on GitHub. So, I would ask that everyone be patient with each other and give the others the benefit of the doubt here. The people participating here care enough about the CNCF to give their time to try and improve things.

I also hope that we can appreciate that some views around diversity are strongly held by some folks where passion comes out.

An aspect I would like everyone to be aware of is the diversity of the CNCF contributors. A super majority of whom are outside of North America. Approximately 1/3 are outside of North America and Europe. In the past 5 years I've had an opportunity to work for an Asian company and a European company while living in North America. Working with others from different regions has been educational. For my own education into the views and beliefs of folks in regions outside of my own I spend regular time reading about them. The CNCF is incredible diverse along many axis.

I don't mean to discount the diverse groups we often talk about with women, LGBTQ+, those of color, etc. I just want to bring to our awareness the amount of diversity we have as a whole. Living in North America and having community meetings in North American friendly time zones I have found it's easy to miss out on the rest of the world. To consider that everyone involved is not like us in North America or the region you might be in.

Another axis to consider is belief system. I personally know people in the CNCF community who have shared with me that they are atheists, Muslims, Christians (including conservative ones), Hindus, Jewish, and more. On the political system we have to consider that we have people who align with various groups in systems that don't clearly map to each other. We can't view people as being like Democrats or Republicans in the US (of which we have active people in both groups).

With all this diversity we are able to come together to work on CNCF projects.

The CNCF's mission:

> The Foundation’s mission is to make cloud native computing ubiquitous. 

A diverse group of people from around the world, hailing from different countries, with different religious, political, and ideological beliefs come together for this mission. The code of conduct can help us do that. We need to keep this wide diversity in mind when creating, updating, and executing on it.

**Reply by endocrimes** [reply]
As a contributor outside of North America, who has worked with teams from around the world, and who has spent a _monumental_ amount of my evenings working with people in open source (and repeatedly had my career be complicated by the fact that I do not have a US passport), I'm very much aware of how much we center North America in a lot of discussions (it can be _incredibly_ frustrating when everyone interprets discussions as being about the US or American politics, and heartbreaking to see friends and coworkers from further east be constantly rejected for visas or passed over for work).

It's because of that level of diversity that we need to define boundaries - and again why I tend to bias towards a principle of “creating and upholding an environment of mutual respect among all people who contribute to the project in any way” - which is one that the current definition shares. A code of conduct is a community saying "this is how we wish to work together, and the _bare minimum_ standards we require people to have for each other" - and by participating, you are agreeing to that principle of mutual respect.

That principle of mutual respect is one that requires saying that tolerance is mutual agreement, not a moral axiom that the world shares - because we have a swath of moral systems.

Which comes back to my earlier point:

> There's an important distinction to be mad between a statement that some do not like, vs one that does harm to members of the community...

When a statement harms others (or has the potential to do so) it shows that the principle of mutual respect has been broken and there are always going to potentially be consequences. Regardless of where that statement comes from.

**Comment by endocrimes** [comment]
An important facet of the code of conduct (or more common enforcement practice) that I don't think we currently cover in any of our documentation is that of understanding power balances, and the difference between "punching down" and "expressing frustration at the decisions of leadership".

It's a fairly common problem that project CoCCs have to deal with, and it's "hard" to document effectively. 

As a concrete but non-real example (note, I am trying incredibly hard to not reference any specific example here to avoid making references to the Kubernetes community):

- A SIG Lead makes a decision to change a meeting time to better fit their schedule, but makes it worse for cross-timezone collaboration.
- A member of the SIG expresses frustration at the Lead for making the meeting time less inclusive and feels like they're being excluded from the SIG. As part of this, they maybe aren't as polite as would be ideal.

There are a few things that could happen. The SIG Lead could:
a) respond emotionally and escalate the situation
b) reach out to the person privately and discuss the change, and find a calm resolution
c) reach out to the CoCC and they can help come to (b)

There are cases where the response could be to reach out to the CoCC with a view to exclude someone based on that frustration - and it's something we should explicitly frown upon unless something is _particularly_ vitriolic. If people feel like they cannot effect change it is expected for them to express that frustration - especially in open source where people tend to care deeply.

on the flip side, if (A) occurs, the SIG Lead potentially risks breaking the code of conduct and requiring the CoCC to help restore community trust.

**Comment by joannalee333** [comment]
In response to comments on GitHub and feedback I've received via other channels, I've modified the scope language in the existing PR to read as follows:

"This code of conduct applies:

*within project and community spaces,
*in other spaces when an individual's words or actions are directed at a CNCF project or the CNCF community, and
*when an individual participant in the CNCF community is directly harassing another participant."
